### PR TITLE
32bit Teensy HAL

### PR DESCRIPTION
**Pull request by teemuatlut**
Add Teensy3.5 and Teensy3.6 to the Marlin HAL.

https://travis-ci.org/teemuatlut/Marlin/branches Teensy-HAL-merged
Travis is updated to run almost all of the checks on Teensy as well. I had to install a lot of utilities and libraries to get it all to compile automated. The Teensyduino from PJRC couldn't be used because of the GUI installer. The whole Travis check runs for about 14mins and should compile fine with both AVR and Teensy.

I've included some simple fixes for other platforms as well. The marlin serial had been moved to AVR HAL and the path needed an update. The Print Counter I took from the current RCBugFix.

I've made a new servo library that is inherited from the core, and just adds the few required additional methods. I saw no need to copy the whole library just to add a little something.

Added a waiting check for Marlin serial since the Teensy needed the while(!Serial) check to print properly on startup.

There was some really weird behavior that I saw with the commit that allowed UART and stepper ISRs inside the temperature ISR. The Teensy would completely Freeze and Windows wouldn't properly recognize it again. It also froze whatever serial communication program I was using at the time. I don't know why it was doing this but maybe someone else can look into it as well.

I'll try to add comments around to explain changes but I'd like to see others review and possibly even try it out, but I think this might be close to a good starting point.

### CODE REVIEW

**Comment by Roxy-3D** [comment]
> There was some really weird behavior that I saw with the commit that allowed UART and stepper ISRs inside the temperature ISR. The Teensy would completely Freeze and Windows wouldn't properly recognize it again. It also froze whatever serial communication program I was using at the time. I don't know why it was doing this but maybe someone else can look into it as well.

I wonder if the 'right' thing to do is take the UART stuff out of the 32-bit temperature ISR initially.  On the 32-bit processors we have plenty of speed, so we don't need the extra checks there.   And we probably want to re-think the stepper motor movements in the Temperature ISR for 32-Bit.  If we have an extra timer available, that would let us do that.  The BabyStepping is probably what you are talking about being in the Temperature ISR.   With a separate timer task running on its own timer,  this could be simplified.

**Comment by teemuatlut** [comment]
I would agree that it could be taken out if it is one the AVR 'hacks'.

I haven't really looked into it, but I think there are interrupt priorities in the ARM processors. Maybe that could be something we could make use of?

**Comment by p3p** [comment]
In my build I need to override the cli and sei calls to "fix" the temperature ISR, I haven't met any issue with doing this after quite a lot of printing, (having appropriate interrupt priorities) but haven't had time to research the root cause, I use interrupts for so much DMA that I assumed it may been specific to my HAL implementation.

**Comment by Roxy-3D** [comment]
> I use interrupts for so much DMA that I assumed it may been specific to my HAL implementation.

@p3p  You  really mean you use interrupts for so much data transfer....   Not literally DMA, right?

**Comment by p3p** [comment]
I guess it would make more sense to word it like that, I have DMA's set up for Serial, ADC and SPI so along with the usual ARM system interrupts there is a lot going on.

File: Marlin/Conditionals_post.h [hunk 1]
```diff
@@ -824,4 +824,12 @@
     #define FMOD(x, y) fmodf(x, y)
   #endif
 
+  #if defined(TEENSYDUINO)
```
Comment (teemuatlut) [inline Marlin/Conditionals_post.h]: Teensy core changes the min/max implementation in a way that we cannot use it. This changes it to the way AVR core does it.

File: Marlin/src/HAL/HAL_AVR/MarlinSerial.h [hunk 2]
```diff
@@ -170,6 +170,7 @@ class MarlinSerial { //: public Stream
     static void println(unsigned long, int = DEC);
     static void println(double, int = 2);
     static void println(void);
+    operator bool() { return true; }
```
Comment (teemuatlut) [inline Marlin/src/HAL/HAL_AVR/MarlinSerial.h]: Needed for while(!Serial) wait loop when compiling with AVR HAL.

**Comment by thinkyhead** [comment]
I'll try to rebase the branch and then review and merge this later on. The rebase may take a little while, as my branch is about 2 weeks stale. I'll see about recording that as a workflow demo.

**Comment by Roxy-3D** [comment]
@teemuatlut   Thank you for the changes and advice!   I just wish you were working with a LPC17868 processor.   But any progress forward is noticed and appreciated!!!

**Comment by teemuatlut** [comment]
@Roxy-3D  I'd love to, but I don't have one and am currently not in a position to make too many new purchases.

> If we have an extra timer available, that would let us do that. The BabyStepping is probably what you are talking about being in the Temperature ISR. With a separate timer task running on its own timer, this could be simplified.

Not about the babystepping I think. This is the [PR](https://github.com/MarlinFirmware/Marlin/pull/5818/files) and [commit](https://github.com/MarlinFirmware/Marlin/commit/271ced73413ebf8d662cce2a582ac9b795104a8c) that added the cli() and sei() lines.
Could you open up your thoughts about the timers a bit more? In the Teensy the millis are from the systick timer, then there are two others that are in use; FlexTimerModule0 for stepper and FTM1 for temperature. No prioritization as of yet.

**Comment by thinkyhead** [comment]
Ok, I've rebased the base branch `32-Bit-RCBugFix-new` onto the latest `RCBugFix`. So, see if you're able to rebase this onto `32-Bit-RCBugFix-new`. Let me know if you need assistance. I'm getting pretty good at all the conflict-resolution tricks.

**Comment by teemuatlut** [comment]
I'll look into it this weekend when I have the time.

**Pull request closed without merging**